### ISSUE TITLE
为什么在一个for 循环中依次训练多个模型，每loop 一次GPU内存都会增加几百M直到oom？

### ISSUE THREAD

**Issue by jeffzhengye** [PaddlePaddle/Paddle#32627]
所有模型定义都在for 循环体内，GPU内存增加是什么原因？

代码如下

`def train_high_level():
    if is_test_run:
        global train_df, epochs
        train_df = train_df.iloc[:50]
        epochs = 1
    folds = StratifiedKFold(n_splits=CFG['fold_num'], shuffle=True, random_state=CFG['seed']) \
        .split(np.arange(train_df.shape[0]), train_df.label.values)  # 五折交叉验证
    only_folds = global_config['only_folds']

    best_values = []
    for fold, (trn_idx, val_idx) in enumerate(folds):
        if only_folds and fold not in only_folds:  # not empty
            continue  # skip if not empty and not in
        logger.info(f'training fold {fold} with only_fold setting: {only_folds}')
        logger.info('current split is: len= {}, top 10 is: {}'.format(len(trn_idx), trn_idx))
        train = train_df.loc[trn_idx]
        val = train_df.loc[val_idx]

        if support_sen_type == 0:
            train_set = MyDataset(train)
            val_set = MyDataset(val)
        elif support_sen_type in [1, 2]:
            train_set = MostSimilarDataset(train, sel_type=support_sen_type, top_k=top_k, train=True,
                                           random_concat=random_concat)
            val_set = MostSimilarDataset(val, sel_type=support_sen_type, top_k=top_k, train=False)

        train_set = MapDataset(train_set)
        train_set.map(train_trans_func, batched=False)

        val_set = MapDataset(val_set)
        val_set.map(train_trans_func, batched=False)

        train_batch_sampler = paddle.io.BatchSampler(
            train_set, batch_size=batch_size, shuffle=True)

        train_data_loader = paddle.io.DataLoader(
            dataset=train_set,
            batch_sampler=train_batch_sampler,
            collate_fn=batchify_fn,
            num_workers=num_workers,
            # batch_size=batch_size,
            return_list=True)
        val_data_loader = paddle.io.DataLoader(
            dataset=val_set,
            collate_fn=batchify_fn,
            batch_size=batch_size,
            return_list=True)
        if bert_type == 1:
            model = ErnieForSequenceClassificationQ4.from_pretrained(MODEL_NAME)
        elif bert_type == 0:
            model = BertForSequenceClassificationQ4.from_pretrained(MODEL_NAME, dropout=dropout)
        elif bert_type == 2:
            model = RobertaForSequenceClassificationQ4.from_pretrained(MODEL_NAME)
        # model = BertForSequenceClassificationQ4(bert=pretrained)
        pmodel = paddle.Model(model)

        num_training_steps = len(train_data_loader) * epochs
        lr_scheduler = ppnlp.transformers.CosineDecayWithWarmup(learning_rate, num_training_steps, warmup_proportion)
        optimizer = paddle.optimizer.AdamW(
            learning_rate=lr_scheduler,
            parameters=model.parameters(),
            weight_decay=weight_decay)

        # Defines loss and metric.
        criterion = paddle.nn.CrossEntropyLoss()
        metric = paddle.metric.Accuracy()
        pmodel.prepare(optimizer, criterion, metric)

        earlystop = paddle.callbacks.EarlyStopping('acc', mode='max', patience=3, verbose=1, save_best_model=True)
        # Starts training and evaluating.
        logger.info('eval fold {} with {} examples'.format(fold, len(val_set)))
        pmodel.fit(train_data_loader, eval_data=val_data_loader, epochs=epochs, verbose=1,
                   save_dir=f'./{model_outputs}{fold}',
                   num_workers=10,
                   callbacks=[earlystop])
        from mmrc.utils.eval_utils import CVRecorder
        recorder = CVRecorder(fold, earlystop.best_value, train_idx=trn_idx, val_idx=val_idx, fold_num=fold_num)
        recorder.save(f'./{model_outputs}{fold}')
        restored = CVRecorder.load(f'./{model_outputs}{fold}')
        if restored is not None:
            logger.debug('pickled and restored with acc ={}'.format(restored.acc))

        best_values.append(earlystop.best_value)
        logger.info('fold {} best acc={}'.format(fold, earlystop.best_value))
        del pmodel, model, earlystop, optimizer, train_data_loader, val_data_loader
        logger.info('best values, mean={}, all={}'.format(np.mean(best_values), best_values))`

**Comment by zhiqiu** [comment]
hi，我们怀疑是某些model占用的显存没有被释放，请问有复现方法吗？

**Comment by jeffzhengye** [comment]
> hi，我们怀疑是某些model占用的显存没有被释放，请问有复现方法吗？

code 全部在这个for 循环里面啊。现在是没有理由不释放啊？
其它code 就没有跟paddle 训练相关的，
如果你们要有经验，应该一眼可以看出code问题在哪里啊，不然估计就是paddle 本身bug了

**Comment by zhiqiu** [comment]
> > hi，我们怀疑是某些model占用的显存没有被释放，请问有复现方法吗？
> 
> code 全部在这个for 循环里面啊。现在是没有理由不释放啊？
> 其它code 就没有跟paddle 训练相关的，
> 如果你们要有经验，应该一眼可以看出code问题在哪里啊，不然估计就是paddle 本身bug了

不好意思，可能从现在的代码还不能确定问题的原因。您方便提供下如何跑的方法吗，我们可能需要用您这份代码运行和调试一下才方便定位具体原因。

**Comment by jeffzhengye** [comment]
> > > hi，我们怀疑是某些model占用的显存没有被释放，请问有复现方法吗？
> > 
> > 
> > code 全部在这个for 循环里面啊。现在是没有理由不释放啊？
> > 其它code 就没有跟paddle 训练相关的，
> > 如果你们要有经验，应该一眼可以看出code问题在哪里啊，不然估计就是paddle 本身bug了
> 
> 不好意思，可能从现在的代码还不能确定问题的原因。您方便提供下如何跑的方法吗，我们可能需要用您这份代码运行和调试一下才方便定位具体原因。

具体可以看工程源码：https://github.com/jeffzhengye/mmrc/tree/main

具体是在训练的时候出问题，运行下面的命令观察gpu内存就可以复现：
·CUDA_VISIBLE_DEVICES=0 python main.py -c configs/haihua.yml -o Global.additional_tag='cosine_aug_norandom' Global.only_folds="[0,1]" Global.num_workers=10 Global.train_bs=20 Global.top_k=5 Global.max_len=256 Global.fold_num=10 Global.epochs=10 Global.support_sen_type=0 Global.random_concat=false Global.is_train=true Global.is_eval=false Global.is_test_run=true·

Global.is_test_run 这个参数就是截取前面的数据测试用的。

**Comment by jeffzhengye** [comment]
@zhiqiu 忘记上传一些问题，现在应该可以直接运行了

**Comment by zhiqiu** [comment]
> @zhiqiu 忘记上传一些问题，现在应该可以直接运行了

用这个代码运行，报错了：[Errno 2] No such file or directory: 'train.json_cosine.csv'

**Comment by jeffzhengye** [comment]
> f

要先解压数据 data.tgz @zhiqiu

**Comment by zhiqiu** [comment]
> > f
> 
> 要先解压数据 data.tgz @zhiqiu

我在16G的V100机器上，bs=20第一个batch的时候就挂了。调成bs=10后，能够正常训练完。

**Comment by jeffzhengye** [comment]
> > > f
> > 
> > 
> > 要先解压数据 data.tgz @zhiqiu
> 
> 我在16G的V100机器上，bs=20第一个batch的时候就挂了。调成bs=10后，能够正常训练完。

我的问题是每训练一个fold（for loop中），GPU内存都要增大几百M大概。
可能需要你观察下内存，

一般我们会让batch_size 尽可能大。
如果bs=10的话，应该每次训练GPU内存还是会增，但是还不到OOV的程度。

**Comment by jeffzhengye** [comment]
@zhiqiu 看出问题了没有？

**Comment by jeffzhengye** [comment]
@zhiqiu

**Comment by weihongwei-zg** [comment]
同样的问题

**Comment by AndPuQing** [comment]
同样问题

**Comment by AndPuQing** [comment]
@zhiqiu  #32927 类似问题

**Comment by zhiqiu** [comment]
hi，非常抱歉许久后才回复您。@jeffzhengye 经过分析，这个问题最近有了一些新的进展，目前定位到该问题的原因是：您在for loop中定义了不同的model，然后在训练后调用了del model。但是python的gc机制并不是立即回收的，所以这个model在执行del之后可能还未释放显存。

解决方法，在del model之后加入gc.collect()，示例如下：

```
for i in range(10):
    model = Model()
    ...
    model.fit()
    ...
    del model
    import gc
    gc.collect()
```

**Comment by jeffzhengye** [comment]
@zhiqiu  这个还是解决不了gpu 内存泄漏的问题。

今天还发现个奇怪的泄露。只要从文件中恢复参数，继续训练就多用1G多内存。

elif args.resume:
        resume_file = os.path.join(args.save_dir, 'model_best', 'model_state.pdparams')
        state_dict = paddle.load(resume_file)
        model.set_dict(state_dict)

**Comment by veyron95** [comment]
@jeffzhengye 麻烦介绍下所使用的环境（包括系统，cuda, cudnn, paddle版本等），以及请提供最小复现demo，方便复现和排查
